File: LNCollectionView/LNScrollView/Simulator/LNScrollViewBounceSimulator.m
//
//  LNScrollViewBounceSimulator.m
//  LNCollectionView
//
//  Created by Levison on 7.11.24.
//

#import "LNScrollViewBounceSimulator.h"

@interface LNScrollViewBounceSimulator ()

@property (nonatomic, assign) CGFloat damping;
@property (nonatomic, assign) NSTimeInterval currentTime;
@property (nonatomic, assign) CGFloat velocity_0;
@property (nonatomic, assign) CGFloat offset;
@property (nonatomic, assign) CGFloat targetPosition;

@property (nonatomic, assign) CGFloat position;
@property (nonatomic, assign) CGFloat velocity;

@end

@implementation LNScrollViewBounceSimulator

- (instancetype)initWithPosition:(CGFloat)position velocity:(CGFloat)velocity targetPosition:(CGFloat)targetPosition
{
    self = [super init];
    if (self) {
        self.damping = 10.9;
        self.targetPosition = targetPosition;
        self.offset = position - targetPosition;
        if (fabs(self.offset) < 1.f) {
            self.currentTime = 0.f;
            self.velocity_0 = velocity;
        } else {
            self.currentTime = 1.f/(self.damping + velocity/self.offset);
            self.velocity_0 = self.offset * exp(self.damping * self.currentTime)/self.currentTime;
        }
    }
    return self;
}

- (CGFloat)position {
    return self.targetPosition + self.offset;
}

- (void)accumulate:(NSTimeInterval)during {
    self.currentTime += during;
    self.offset = self.velocity_0 * self.currentTime * exp(- self.damping * self.currentTime);
    self.velocity = self.velocity_0 * exp(- self.damping * self.currentTime);
}

- (BOOL)isFinished
{
    return fabs(self.offset) < 0.1f && fabs(self.velocity) < 0.01;
}

@end

File: Android/app/src/main/java/com/ln/collectionview/scrollview/simulator/LNScrollViewBounceSimulator.kt
package com.ln.collectionview.scrollview.simulator

import android.util.Log

class LNScrollViewBounceSimulator(
    initialPosition: Float,
    initialVelocity: Float,
    targetPosition: Float
) {

    private var damping: Float = 10.9f
    private var currentTime: Float = 0f
    private var velocity0: Float = 0f
    var offset: Float = initialPosition - targetPosition
    private var targetPosition: Float = targetPosition

    var velocity: Float = initialVelocity
        private set

    fun position():Float {
        return targetPosition + offset
    }

    init {
        offset = initialPosition - targetPosition
        if (Math.abs(offset) < 1f) {
            currentTime = 0f
            velocity0 = initialVelocity
        } else {
            currentTime = 1f / (damping + initialVelocity / offset)
            velocity0 = offset * Math.exp((damping * currentTime).toDouble()).toFloat() / currentTime
        }
    }

    fun accumulate(during: Float) {
        currentTime += during
        offset = velocity0 * currentTime * Math.exp(-damping * currentTime.toDouble()).toFloat()
        velocity = velocity0 * Math.exp(-damping * currentTime.toDouble()).toFloat()
    }

    fun isFinished(): Boolean {
        return Math.abs(offset) < 0.1f && Math.abs(velocity) < 0.01f
    }
}
